File: csw-event-impl/src/test/scala/csw/services/event/internal/redis/RedisPubSubTest.scala
package csw.services.event.internal.redis

import akka.actor.ActorSystem
import com.github.sebruck.EmbeddedRedis
import csw.messages.commons.CoordinatedShutdownReasons.TestFinishedReason
import csw.services.event.RedisFactory
import csw.services.event.helpers.RegistrationFactory
import csw.services.event.helpers.TestFutureExt.RichFuture
import csw.services.event.internal.EventServicePubSubTestFramework
import csw.services.event.internal.commons.{EventServiceConnection, Wiring}
import csw.services.location.commons.ClusterAwareSettings
import csw.services.location.scaladsl.LocationServiceFactory
import io.lettuce.core.RedisClient
import org.scalatest.{BeforeAndAfterAll, FunSuite, Matchers}
import redis.embedded.RedisServer

class RedisPubSubTest extends FunSuite with Matchers with BeforeAndAfterAll with EmbeddedRedis {
  private val seedPort        = 3558
  private val redisPort       = 6379
  private val clusterSettings = ClusterAwareSettings.joinLocal(seedPort)
  private val locationService = LocationServiceFactory.withSettings(ClusterAwareSettings.onPort(seedPort))
  private val tcpRegistration = RegistrationFactory.tcp(EventServiceConnection.value, redisPort)
  locationService.register(tcpRegistration).await

  private val redis = RedisServer.builder().setting(s"bind ${clusterSettings.hostname}").port(redisPort).build()

  private implicit val actorSystem: ActorSystem = clusterSettings.system
  private val redisClient                       = RedisClient.create()
  private val wiring                            = new Wiring(actorSystem)
  private val redisFactory                      = new RedisFactory(redisClient, locationService, wiring)
  private val publisher                         = redisFactory.publisher().await
  private val subscriber                        = redisFactory.subscriber().await
  private val framework                         = new EventServicePubSubTestFramework(publisher, subscriber)

  override def beforeAll(): Unit = {
    redis.start()
  }

  override def afterAll(): Unit = {
    redisClient.shutdown()
    wiring.shutdown(TestFinishedReason).await
    redis.stop()
  }

  test("Redis pub sub") {
    framework.pubSub()
  }

  test("Redis independent subscriptions") {
    framework.subscribeIndependently()
  }

  ignore("Redis multiple publish") {
    framework.publishMultiple()
  }

  test("Redis retrieve recently published event on subscription") {
    framework.retrieveRecentlyPublished()
  }

  test("Redis retrieveInvalidEvent") {
    framework.retrieveInvalidEvent()
  }

  test("Redis get") {
    framework.get()
  }

  test("Redis get retrieveInvalidEvent") {
    framework.retrieveInvalidEventOnget()
  }

}

File: csw-event-impl/src/test/java/redis/RedisPubSubTest.java
//package redis;
//
//import akka.actor.ActorSystem;
//import akka.stream.scaladsl.Sink;
//import csw.messages.events.Event;
//import csw.messages.events.EventKey;
//import csw.services.event.JRedisFactory;
//import csw.services.event.RedisFactory;
//import csw.services.event.helpers.Utils;
//import csw.services.event.internal.commons.EventServiceConnection;
//import csw.services.event.internal.commons.Wiring;
//import csw.services.event.javadsl.IEventPublisher;
//import csw.services.event.scaladsl.EventSubscriber;
//import csw.services.location.commons.ClusterAwareSettings;
//import csw.services.location.commons.ClusterSettings;
//import csw.services.location.commons.RegistrationFactory;
//import csw.services.location.models.TcpRegistration;
//import csw.services.location.scaladsl.LocationService;
//import csw.services.location.scaladsl.LocationServiceFactory;
//import io.lettuce.core.RedisClient;
//import org.junit.Test;
//import redis.embedded.RedisServer;
//import scala.collection.immutable.ListSet;
//import scala.collection.immutable.Set;
//import scala.collection.immutable.Set$;
//
//import java.util.concurrent.ExecutionException;
//
//public class RedisPubSubTest {
//    private int seedPort = 3558;
//    private int redisPort = 6379;
//    private ClusterSettings clusterSettings = ClusterAwareSettings.joinLocal(seedPort);
//    private LocationService locationService = LocationServiceFactory.withSettings(ClusterAwareSettings.onPort(seedPort));
//    private TcpRegistration tcpRegistration = RegistrationFactory.tcp(EventServiceConnection.value(), redisPort);
//    private RedisServer redis = RedisServer.builder().setting("bind ${clusterSettings.hostname}").port(redisPort).build();
//    private ActorSystem actorSystem = clusterSettings.system();
//    private RedisClient redisClient = RedisClient.create();
//    private Wiring wiring = new Wiring(actorSystem);
//    private JRedisFactory redisFactory = new JRedisFactory(redisClient, locationService, wiring);
//    private IEventPublisher publisher = redisFactory.publisher().get();
//    private EventSubscriber subscriber = redisFactory.subscriber().get();
//
//    public RedisPubSubTest() throws ExecutionException, InterruptedException {
//        locationService.register(tcpRegistration);
//    }
//    @Test
//    public void shouldBeAbleToPublishAndSubscribeAnEvent() {
//        Event event1      = Utils.makeDistinctEvent(1);
//        EventKey eventKey = event1.eventKey();
//
//        Set<EventKey> set = new ListSet<EventKey>().$plus(eventKey);
//
//        Thread.sleep(2000);
//
//        publisher.publish(event1);
//        Thread.sleep(1000);
//
//        subscription.unsubscribe().await
//        seqF.await shouldBe List(Event.invalidEvent, event1)
//    }
//}
